### ISSUE TITLE
Proposal: Relax instances for Functor combinators; put superclasses on <class>1 to make less-breaking 

### ISSUE THREAD

**Issue by Ericson2314** [haskell/core-libraries-committee#10]
**Implementation** https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4727

The first change makes the `Eq`, `Ord`, `Show`, and `Read` instances for `Sum`, `Product`, and `Compose` match those for `:+:`, `:*:`, and `:.:`. These have the proper flexible contexts that are exactly what the instance needs:

For example, instead of
```haskell
instance (Eq1 f, Eq1 g, Eq a) => Eq (Compose f g a) where
  (==) = eq1
```
we do
```haskell
deriving instance Eq (f (g a)) => Eq (Compose f g a)
```

But, that change alone is rather breaking, because until now `Eq (f a)` and `Eq1 f` (and respectively the other classes and their `*1` equivalents too) are *incomparable* constraints. This has always been an annoyance of working with the `*1` classes, and now it would rear it's head one last time as an pesky migration.

Instead, we give the `*1` classes superclasses, like so:

```haskell
(forall a. Eq a => Eq (f a)) => Eq1 f
```

along with some laws that canonicity is preserved, like:

```haskell
liftEq (==) = (==)
```

and likewise for `*2` classes

```haskell
(forall a. Eq a => Eq1 (f a)) => Eq2 f
```
along with some laws that canonicity is preserved, like:

```haskell
liftEq2 (==) = liftEq1
```

The `*1` classes also have default methods using the `*2` classes where possible.

What this means, as explained in the docs in my implementation, is that `*1` classes really are generations of the regular classes, indicating that the methods can be split into a canonical lifting combined with a canonical inner, with the super class "witnessing" the laws[1] in a fashion.

Circling back to the pragmatics of migrating, note that the superclass means evidence for the old `Sum`, `Product`, and `Compose` instances is (more than) sufficient, so breakage is less likely --- as long no instances are "missing", existing polymorphic code will continue to work.

Breakage can occur when a datatype implements the `*1` class but not the corresponding regular class, but this is almost certainly an oversight. For example, containers made that mistake for `Tree` and `Ord`, which I fixed in https://github.com/haskell/containers/pull/761, but fixing the issue by adding `Ord1` was extremely *un*controversial.

-----

[1]: In fact, someday, when the laws are part of the language and not
only documentation, we might be able to drop the superclass field of the
dictionary by using the laws to recover the superclass in an
instance-agnostic manner, e.g. with a *non*-overloaded function with
type:

```haskell
DictEq1 f -> DictEq a -> DictEq (f a)
```

But I don't wish to get into optomizations now, just demonstrate the
close relationship between the law and the superclass.

**Comment by Ericson2314** [comment]
When I rebased this, I found that `Generically1` was missing `Eq, `Ord`, `Show`, and `Read` instances, even though it had the `*1` versions. Classic!

**Comment by Ericson2314** [comment]
https://github.com/haskellari/some/pull/21 This is an example of an downstream library improvement that is blocked on this change.

`GShow f = forall a. Show (f a)` for all intents and purposes, but we cannot make a superclass to indicate (half of) that because the `GShow` instances for `Sum` and `Product` conflict with the overly-conservative `Show1`-based instances we have today, instead of the strictly more general instances proposed here.

**Comment by gwils** [comment]
The canonicity law looks good to me -- I'm actually surprised that `Eq1` and co didn't already have something like this. The quantified constraint superclass also looks fine and sensible to me.
If we're going to use `FlexibleContexts`-style `Eq (f a)` constraints around the place after this change, what reason do `Eq1` and co still have for existing? I thought their main purpose was to avoid such `FlexibleContexts`-style code.

Do we know how much of hackage this breaks? It looks like there was some effort to run this against `head.hackage` but nobody understood the outcome?

**Comment by Ericson2314** [comment]
>  If we're going to use `FlexibleContexts`-style `Eq (f a)` constraints around the place after this change, what reason do `Eq1` and co still have for existing? I thought their main purpose was to avoid such `FlexibleContexts`-style code.

Besides back-compat (just getting rid of the classes altogether would be far more breaking), the classes are still meaningful if one wants to lift a *non*-canonical function on the inside. I am not sure how often that is done, but it's theoretically useful enough that I don't feel inspired to rid of this.

> Do we know how much of hackage this breaks? It looks like there was some effort to run this against `head.hackage` but nobody understood the outcome?

I will try running it again. I think it was just broken then, but might work now.

**Comment by Bodigrim** [comment]
`Eq1` class is still meaningful, because it allows to check heterogeneous equality like "does `f a` _matches_ `f b`?".

**Comment by gwils** [comment]
Thanks, I now agree that `Eq1` and friends still have reason to exist.
@Ericson2314 did the GHC CI help you with an impact analysis?

**Comment by Bodigrim** [comment]
> along with some laws that canonicity is preserved

Are these laws unconditional or are they expecting that an underlying `(==)` is lawful? (I'm thinking about `instance Ord1 Down` issues)

**Comment by cgibbard** [comment]
This is something I'd also love to see some impact analysis on, but I can imagine most users of the *1 classes being happy about the improvement, even if it might involve knocking out some forgotten instances. I'd also personally be happy about the fixes for `Compose`. So this is a tentative +1 from me, unless we somehow see that the world would be covered in tiny fires.

**Comment by Bodigrim** [comment]
I'm tentatively in favor of this proposal, it reflects my experience with `*1` classes and `QuantifiedContraints`. But we really need an impact assessment to understand what we are voting for.

**Comment by Bodigrim** [comment]
@phadej could you possible please share instructions how to test changes to `base` against Stackage, as you did in https://github.com/haskell/core-libraries-committee/issues/3#issuecomment-953389734 ?

**Comment by phadej** [comment]
> The canonicity law looks good to me -- I'm actually surprised that Eq1 and co didn't already have something like this. The quantified constraint superclass also looks fine and sensible to me.

The `Eq1` class today doesn't have any relationship with `Eq`, thus stating the `liftEq (==) = (==)` law is awkward. Quantified super constraint adds necessary bits so the law always type-checks.

>  the classes are still meaningful if one wants to lift a non-canonical function on the inside. I am not sure how often that is done, but it's theoretically useful enough that I don't feel inspired to rid of this.

E.g. `QuickCheck` `listOf` combinator is very much used, it's a special case of `liftArbitrary`. i also use `liftShowsPrec` quite a lot to show lists of things which don't (or cannot) have good `Show` instance (or have bad one).

> Are these laws unconditional or are they expecting that an underlying (==) is lawful? (I'm thinking about instance Ord1 Down issues)

I'd say it's fair to assume that `(==) :: a -> a -> Bool` is lawful. We can say that for every reflexive, transitive `f`, `liftEq f` is also reflexive and transitive: That is relaxed version as `liftEq f _ _ = True` is trivially reflexive and transitive for any `f`, but otoh it's stricter as `liftEq` implementation may assume that `f` is reflexive and transitive.

> could you possible please share instructions how to test changes to base against Stackage

I'll try to do that asap. (They aren't complicated, but not entirely trivial). A good starting point is to have a patch applied to a version of GHC and base which has stackage snapshot (e.g. 9.0.1 or 8.10.x). Migrating stackage to ghc-head is too much work.

**Comment by Ericson2314** [comment]
> Are these laws unconditional or are they expecting that an underlying (==) is lawful?

I think they need not relay on a sane underlying `(==)`. e.g. `[NaN, Nan] :: [Double]` should not be equal to itself, right?

In fact,
```
liftEq f = and . liftA2 f
liftCompare f = foldr thenCmp EQ . liftA2 f
```
might be valid laws too. But I don't know what the cost of those super-classes would be, and especially `thenCmp` not being associative, so I rather punt on that until later :).

**Comment by phadej** [comment]
There are also wrappers in other core-ish libraries, e.g. `Reverse` and `Backwards`:
- https://hackage.haskell.org/package/transformers-0.6.0.2/docs/Data-Functor-Reverse.html
- https://hackage.haskell.org/package/transformers-0.6.0.2/docs/Control-Applicative-Backwards.html

Would be nice if instances were uniform for these.

**Comment by phadej** [comment]
Does this proposal break or not code like:

```haskell
class Eq1 t => Hashable1 t where ...
instance (Hashable1 f, Hashable1 g) => Hashable1 (Compose f g) where ...
```

I have to think about this, so probably a good idea to explicitly write the answer down in the proposal.

**Comment by Bodigrim** [comment]
@phadej just a gentle reminder about https://github.com/haskell/core-libraries-committee/issues/10#issuecomment-968068641

@Ericson2314 any chance to tests against head.hackage?

**Comment by phadej** [comment]
@Bodigrim as I said, the first step is to have a patch against a GHC version which has a stackage snapshot. GHC `master` and having to use `head.hackage` just adds to much variables, that it's not feasible.

Then the simple approach is to create a cabal package with
- a `cabal.config` (add it to a url of a snapshot e.g. https://www.stackage.org/nightly-2021-12-17/cabal.config) changing
   - `constraints` to `build-depends` 
   - removing `installed` packages
- the `cabal build`
   - some packages are `executable` only. You may remove them or move to `build-tool-depends` (but then you need to figure out the executable names, which usually are the same as package ...)

This approach also ignores flags.

This all can be automated, by
- using YAML snapshot specs from https://github.com/commercialhaskell/stackage-snapshots
   - this would allow to pin flags too.
   - possibly add diagnostics when revisions are made
- using `cabal-install-parsers` to read the index (to figure out which packages are executables, diagnostics mentioned above)

And one general tip is to do 100 packages at the time, by first doing `cabal
build --dry`: Solver in all cabal installs is slow when there are many direct
dependencies (i think it's improved in `master`), as build plans will fail and
you'll need to figure out missing native dependencies or just disable these
packages.  Such metadata would be great to have up front, and the tool would
figure that out directly as here we are not really solving anything.
It might be it's slow because I didn't constraint flags,
so solver tries to flip them.

**Comment by Ericson2314** [comment]
The Hashable instance is fine. See how the Eq1 instance for Compose is unaffected on the patch. I would recommend those packages adjust the definition of Hashable itself analogously, but it's not mandatory.

**Comment by Ericson2314** [comment]
Hmm I thought this stuff was already automated, but I guess not.

**Comment by Bodigrim** [comment]
@Ericson2314 any chance for impact assessment please? I'd like to resolve this one way or another.

**Comment by Ericson2314** [comment]
Sorry, I have not blocked out the chunk of time needed to get through that fairly manual process yet. I will try to bump this forward in my queue of things to do.

**Comment by Bodigrim** [comment]
@Ericson2314 I've created https://github.com/Bodigrim/clc-stackage, which is a meta-package for the majority of package in Stackage Nightly. Essentially all you need to do is to backport your changes to base-4.15, compile GHC from 9.0 branch and run `cabal build -w ghc-9.0.your_patch`. You can further provide fixes for affected packages via local copies, referenced from `cabal.project`. This should give a rough idea how much is broken.

**Comment by Ericson2314** [comment]
Thanks @Bodigrim. I am hoping after https://github.com/haskellfoundation/tech-proposals/pull/27 goes though this process will be quite polished, and then I will give this a shot.

------

Separately, doing
```haskell
git log libraries/base/Data/Functor/Classes.hs
````
turned up https://github.com/ghc/ghc/commit/e0e03d5b9d5cd678f6402534451964d491f16540 which linked https://mail.haskell.org/pipermail/libraries/2015-July/026014.html

It looks like these `*1` classes were *only* added to base based of the data types that use that had these instances. With this (or #35), those data types no longer need those `*1` classes to implement the more basic ones. What that means is that we can invert the dependency and make the modules with `*1` classes depend on those data types instead.

This makes me also want to just remove the `Ord1` classes from `base` entirely! So
 - #35 is the plan for base
 - Create new `functor-classes` package

The question remains, should this be done right away, or as a successive step in a longer deprecation cycle?

I propose that https://gitlab.haskell.org/ghc/ghc/-/issues/20647 / https://discourse.haskell.org/t/re-pre-pre-hftp-decoupling-base-and-ghc/4269/3 gives us an alternative.

1. 
    - Do this proposal as originally written
    - Move `base` to `ghc-base` in the GHC repo, with `base` a trivial package (in another repo) that reexports `ghc-base`.

2. Fork out `functor-classes`, making it depend on `ghc-base`, and make `base` reexport `functor-classes` too.

3. After a deprecation cycle, get rid of the `functor-classes` re-export.

This means we still get a smooth migration, but the benefit of making obscure stuff out of tree immediately. This woks best if https://gitlab.haskell.org/ghc/ghc/-/issues/4879 / https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0134-deprecating-exports-proposal.rst is finally implemented, so those reexports can be deprecated immediately.

**Comment by Icelandjack** [comment]
Several instances were omitted from `Generically1`, both to simplify its inclusion and in an effort to make `deriving X via Generically(1) D` match  `deriving stock X` as closely as possible. The `Read(1)` and `Show(1)` instances were not part of that selected instances. I will make an issue proposing the addition to `Eq` and `Ord` but they are subtly different. With `data D = D deriving stock Eq` we get `undefined == D` = `undefined`. However deriving via `Generically1 D` gives `undefined == D` = `True`.

**Comment by Bodigrim** [comment]
I tried to run a quick-and-dirty impact assessment. Most of the packages are fine (indeed, it is rare to define `instance Eq1 f` without defining `instance Eq (f a)` nearby), but `transformers-compat` is seriously affected, because it contains a Generic mechanism to derive `Eq1` and friends:

```
Building library for transformers-compat-0.7.1..
[1 of 4] Compiling Control.Monad.Trans.Instances
[2 of 4] Compiling Data.Functor.Classes.Generic.Internal

generics/Data/Functor/Classes/Generic/Internal.hs:819:10: error:
    • Could not deduce (Eq (FunctorClassesDefault f a))
        arising from the superclasses of an instance declaration
      from the context: (GEq1 NonV4 (Rep1 f), Generic1 f)
        bound by the instance declaration
        at generics/Data/Functor/Classes/Generic/Internal.hs:819:10-75
      or from: Eq a
        bound by a quantified context
        at generics/Data/Functor/Classes/Generic/Internal.hs:1:1
    • In the instance declaration for ‘Eq1 (FunctorClassesDefault f)’
    |
819 | instance (GEq1 NonV4 (Rep1 f), Generic1 f) => Eq1 (FunctorClassesDefault f) where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

@RyanGlScott what do you think?

**Comment by RyanGlScott** [comment]
Hm. `FunctorClassesDefault` is basically `Generically1`, although it provides more instances than `Generically1` does (`Read1` and `Show1`). I tried to see how `Generically1`'s `Eq1` instance works, and it's simpler than expected:

```hs
instance (Generic1 f, Eq (Rep1 f a)) => Eq (Generically1 f a)
```

I say "simpler than expected" because I think you would achieve pretty counterintuitive results if you tried to actually derive an `Eq` instance via `Generically1`. Here is an example which illustrates this:

```hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE DerivingVia #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE StandaloneKindSignatures #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE UndecidableInstances #-}
module Foo where

import Data.Functor.Classes
import Data.Kind
import GHC.Generics

type    Generically1 :: forall k. (k -> Type) -> (k -> Type)
newtype Generically1 f a where
  Generically1 :: forall {k} f a. f a -> Generically1 @k f a

instance (Generic1 f, Eq (Rep1 f a)) => Eq (Generically1 f a) where
   Generically1 x == Generically1 y = from1 x == from1 y
   Generically1 x /= Generically1 y = from1 x /= from1 y

instance (Generic1 f, Eq1 (Rep1 f)) => Eq1 (Generically1 f) where
  liftEq :: (a1 -> a2 -> Bool) -> (Generically1 f a1 -> Generically1 f a2 -> Bool)
  liftEq (===) (Generically1 as1) (Generically1 as2) = liftEq (===) (from1 as1) (from1 as2)

type T :: (Type -> Type) -> (Type -> Type) -> Type -> Type
data T f g a = MkT (f (g a))
  deriving stock Generic1
  deriving Eq via Generically1 (T f g) a
```

Surprisingly, this doesn't work:

```
[1 of 1] Compiling Foo              ( Foo.hs, interpreted )

Foo.hs:33:12: error:
    • No instance for (Eq (f (Rec1 g a)))
        arising from the 'deriving' clause of a data type declaration
      Possible fix:
        use a standalone 'deriving instance' declaration,
          so you can specify the instance context yourself
    • When deriving the instance for (Eq (T f g a))
   |
33 |   deriving Eq via Generically1 (T f g) a
   |            ^^
```

In order to make this work, you have to carefully use `StandaloneDeriving` to give it the right instance context:

```hs
deriving via Generically1 (T f g) a instance (Eq (f (Rec1 g a)), Functor f) => Eq (T f g a)
```

Note that the instance context mentions leaks the generic representation type `Rec1`, which is usually considered a datatype-generic programming faux pas. The instance would probably work if you used it, but I could think of more straightforward ways to define this instance.

-----

Returning back to the topic of `FunctionClassesDefault`, I think it would suffice to give it an `Eq` instance like this:

```hs
instance (GEq (Rep1 f a), Generic1 f) => Eq (FunctorClassesDefault f a) where
  FunctorClassesDefault x == FunctorClassesDefault y = eqDefault x y

class GEq a where
  geq :: a -> a -> Bool

-- Provide all of the usual instances for GEq

eqDefault :: (GEq (Rep1 f a), Generic1 f) => f a -> f a -> Bool
eqDefault m n = geq (from1 m) (from1 n)
```

And then give `GEq1` a quantified superclass on sufficiently recent versions of GHC:

```hs
class (forall a. Eq a => GEq (t a)) => GEq1 v t where
  -- ...
```

The trickiest part is coming up with the `GEq` instance for `(:.:)`. Here is the best thing I have come up with:

```hs
instance (Eq1 f, GEq (g p)) => GEq ((f :.: g) p) where
  geq (Comp1 m) (Comp1 n) = liftEq geq m n
```

With this approach, if you were to derive an `Eq` instance for `T` via `FunctorClassesDefault`, you'd end up with:

```hs
instance (Eq1 f, Eq1 g, Eq a, Functor f) => Eq (T f g a)
```

This seems pretty consistent with the behavior of the derived `Eq1` instances, so I'd be happy with this. When I have some more time, I can port over all of `transformers-compat`'s generic machinery to use this approach. Luckily, it should be backwards-compatible to do so.

**Comment by Bodigrim** [comment]
Thanks for investigation and analysis, @RyanGlScott. 

@Ericson2314 the progress of the proposal mostly depends on updating `transformers-compat` in a backwards-compatible way. Maybe you can help Ryan with it?..

**Comment by Ericson2314** [comment]
Thank you both very much!!!! I'll happily help out @RyanGlScott with that.

**Comment by RyanGlScott** [comment]
@Ericson2314, I've opened ekmett/transformers-compat#54, which I believe should future-proof `transformers-compat` against this change to the superclasses of `Eq1` and friends. I don't have a local build of https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4727 handy, so would you be willing to test that patch? `cabal test all` should suffice.

**Comment by Bodigrim** [comment]
> For example, instead of
> 
> ```haskell
> instance (Eq1 f, Eq1 g, Eq a) => Eq (Compose f g a) where
>   (==) = eq1
> ```
> 
> we do
> 
> ```haskell
> deriving instance Eq (f (g a)) => Eq (Compose f g a)
> ```

That would be a very welcome change, I think. The reason is that GHC `Compose f g a` is kind-polymorphic:
```
> import Data.Functor.Compose
> :k Compose
Compose :: (k -> *) -> (k1 -> k) -> k1 -> *
```

But you cannot have `Eq1 g` for `g :: k1 -> k`, because `Eq1` demands `g :: * -> *`. It gets especially embarassing when `a` is a phantom type: it does not affect `Eq` instance at all, and still one cannot define it.

**Comment by RyanGlScott** [comment]
After some thought, I've realized that the strange `Eq (f (Rec1 g a))` constraint that arises when trying to derive an `Eq` instance via `Generically1` (as seen in https://github.com/haskell/core-libraries-committee/issues/10#issuecomment-1166302780) isn't really the fault of this proposal, but rather due to the unusual way that `(:.:)` is used derived `Generic1` instances. #75 is a proposal to change the conventions surrounding `(:.:)`. If I were to derive an `Eq` instance via `Generically1` using the conventions suggested in that proposal:

<details>

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE DerivingVia #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE StandaloneKindSignatures #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
module Foo where

import Data.Functor.Classes
import Data.Kind
import GHC.Generics

type    Generically1 :: forall k. (k -> Type) -> (k -> Type)
newtype Generically1 f a where
  Generically1 :: forall {k} f a. f a -> Generically1 @k f a

instance (Generic1 f, Eq (Rep1 f a)) => Eq (Generically1 f a) where
   Generically1 x == Generically1 y = from1 x == from1 y
   Generically1 x /= Generically1 y = from1 x /= from1 y

instance (Generic1 f, Eq1 (Rep1 f)) => Eq1 (Generically1 f) where
  liftEq :: (a1 -> a2 -> Bool) -> (Generically1 f a1 -> Generically1 f a2 -> Bool)
  liftEq (===) (Generically1 as1) (Generically1 as2) = liftEq (===) (from1 as1) (from1 as2)

infixr 7 :$:
newtype (:$:) (f :: k2 -> Type) (g :: k1 -> k2) (p :: k1) =
    Appl1 { unAppl1 :: f (g p) }
  deriving Eq

type T :: (Type -> Type) -> (Type -> Type) -> Type -> Type
data T f g a = MkT (f (g a))
  deriving Eq via Generically1 (T f g) a

instance Generic1 (T f g) where
  type Rep1 (T f g) =
    D1 ('MetaData "T" "Foo" "main" 'False)
       (C1 ('MetaCons "MkT" 'PrefixI 'False)
           (S1 ('MetaSel 'Nothing 'NoSourceUnpackedness 'NoSourceStrictness 'DecidedLazy)
               ((Par1 :$: f) :$: g)))
  from1 x = M1 (case x of MkT g -> M1 (M1 (Appl1 (Appl1 (Par1 g)))))
  to1 (M1 x) = case x of M1 (M1 g) -> MkT (unPar1 (unAppl1 (unAppl1 g)))
```
</details>

Then the derived `Eq` instance for `T` becomes:

```hs
instance Eq (f (g a)) => Eq (T f g a)
```

Which is exactly what we want in this proposal. Therefore, I'm not too bothered by the strangeness observed in https://github.com/haskell/core-libraries-committee/issues/10#issuecomment-1166302780. That's due to a quirk in the way `deriving Generic1` works, not this proposal.

**Comment by RyanGlScott** [comment]
Also, thank you to @Bodigrim for testing the patch in ekmett/transformers-compat#54. I've released [`transformers-compat-0.7.2`](https://hackage.haskell.org/package/transformers-compat-0.7.2) to Hackage with these changes.

**Comment by Bodigrim** [comment]
Thanks for your work and quick response, @RyanGlScott, much appreciated. 

I made some further progress with impact assessment. Several packages are easy to fix:
```cabal
source-repository-package
  type: git
  location: https://github.com/Bodigrim/streaming.git

source-repository-package
  type: git
  location: https://github.com/Bodigrim/free.git

source-repository-package
  type: git
  location: https://github.com/Bodigrim/nonemptymap.git
```

The next stumbling block is `text-show`. It introduces a hierarchy of `TextShow`, `TextShow1` and `TextShow2`, which is now subtly broken. Not sure what's the best course of actions here.

```
src/TextShow/FromStringTextShow.hs:175:10: error:
    • Could not deduce (TextShow a)
        arising from the superclasses of an instance declaration
      from the context: Show a
        bound by a quantified context
        at src/TextShow/FromStringTextShow.hs:1:1
      Possible fix:
        add (TextShow a) to the context of a quantified context
    • In the instance declaration for ‘Show1 FromTextShow’
    |
175 | instance Show1 FromTextShow where
    |          ^^^^^^^^^^^^^^^^^^

src/TextShow/FromStringTextShow.hs:328:10: error:
    • Could not deduce (TextShow a)
        arising from the superclasses of an instance declaration
      from the context: TextShow1 f
        bound by the instance declaration
        at src/TextShow/FromStringTextShow.hs:328:10-47
      or from: Show a
        bound by a quantified context
        at src/TextShow/FromStringTextShow.hs:1:1
      Possible fix:
        add (TextShow a) to the context of a quantified context
    • In the instance declaration for ‘Show1 (FromTextShow1 f)’
    |
328 | instance TextShow1 f => Show1 (FromTextShow1 f) where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/TextShow/FromStringTextShow.hs:502:10: error:
    • Could not deduce (TextShow a1)
        arising from the superclasses of an instance declaration
      from the context: (TextShow2 f, TextShow a)
        bound by the instance declaration
        at src/TextShow/FromStringTextShow.hs:502:10-63
      or from: Show a1
        bound by a quantified context
        at src/TextShow/FromStringTextShow.hs:1:1
      Possible fix:
        add (TextShow a1) to the context of a quantified context
    • In the instance declaration for ‘Show1 (FromTextShow2 f a)’
    |
502 | instance (TextShow2 f, TextShow a) => Show1 (FromTextShow2 f a) where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/TextShow/FromStringTextShow.hs:509:10: error:
    • Could not deduce (TextShow a)
        arising from the superclasses of an instance declaration
      from the context: TextShow2 f
        bound by the instance declaration
        at src/TextShow/FromStringTextShow.hs:509:10-47
      or from: (Show a, Show b)
        bound by a quantified context
        at src/TextShow/FromStringTextShow.hs:1:1
      Possible fix:
        add (TextShow a) to the context of a quantified context
    • In the instance declaration for ‘Show2 (FromTextShow2 f)’
    |
509 | instance TextShow2 f => Show2 (FromTextShow2 f) where
    |          ^^^^^^^^^^^^^^^^
```

**Comment by Bodigrim** [comment]
The closest I can get is https://github.com/Bodigrim/text-show/commit/e4fc3c759a9e601063882c303c1bb6c2b4e9bd4c, but `forall a. Show a => TextShow a` is a bit weird. @RyanGlScott, sorry to bother you, may I ask for your opinion on this please?

**Comment by Ericson2314** [comment]
This is interesting. I think we have two options here

1. Delete `Show1 FromTextShow` for being incorrect now, because having a `Show a` is *not* sufficient to have a `Show (FromTextShow a)`: one instead needs a `TextShow a` and that is indeed the very point of this class.

2. Add a `Show a` superclass to `TextShow a`, requiring every `TextShow` instance to have a matching `Show` instance. That would work, but is against the sort of usage where one is trying to avoid `Show` entirely.

**Comment by Ericson2314** [comment]
```haskell
instance (forall a. Show a => TextShow a) => Show1 FromTextShow
```
is tantamount to option 1, because that constraint will never be satisfied.

**Comment by RyanGlScott** [comment]
This is a tricky one, and I don't yet have a good idea of how to adapt this code.

`FromTextShow` is a newtype used for deriving `Show` instances via `TextShow`. There is a similar relationship between `FromTextShow1`/`Show1`/`TextShow1` and `FromTextShow2`/`Show2`/`TextShow2`. I'll focus on `FromTextShow1`'s instances for now:

```hs
instance (TextShow1 f, TextShow a) => Show (FromTextShow1 f a)
instance TextShow1 f => Show1 (FromTextShow1 f)
```

When `Show1` has a quantified superclass, GHC will reject the `Show1` instance for `FromTextShow1`:

```
src/TextShow/FromStringTextShow.hs:355:10: error:
    • Could not deduce (TextShow a)
        arising from the superclasses of an instance declaration
      from the context: TextShow1 f
        bound by the instance declaration
        at src/TextShow/FromStringTextShow.hs:355:10-47
      or from: Show a
        bound by a quantified context
        at src/TextShow/FromStringTextShow.hs:1:1
      Possible fix:
        add (TextShow a) to the context of a quantified context
    • In the instance declaration for ‘Show1 (FromTextShow1 f)’
    |
355 | instance TextShow1 f => Show1 (FromTextShow1 f) where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

And indeed, the `Show` instance has `TextShow a` in its instance context, but the `Show1` instance doesn't satisfy that constraint. My first inclination was to change the context in the `Show` instance:

```hs
instance TextShow (f a) => Show (FromTextShow1 f a)
```

That changes the error message to:

```
src/TextShow/FromStringTextShow.hs:358:10: error:
    • Could not deduce (TextShow (f a))
        arising from the superclasses of an instance declaration
      from the context: TextShow1 f
        bound by the instance declaration
        at src/TextShow/FromStringTextShow.hs:358:10-47
      or from: Show a
        bound by a quantified context
        at src/TextShow/FromStringTextShow.hs:1:1
    • In the instance declaration for ‘Show1 (FromTextShow1 f)’
    |
358 | instance TextShow1 f => Show1 (FromTextShow1 f) where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

OK, fair enough. Next, I gave `TextShow1` a quantified superclass:

```hs
class (forall a. TextShow a => TextShow (f a)) => TextShow1 f where
  -- ...
``` 

But that just changes the error message back to what it was originally!

```
src/TextShow/FromStringTextShow.hs:358:10: error:
    • Could not deduce (TextShow a)
        arising from the superclasses of an instance declaration
      from the context: TextShow1 f
        bound by the instance declaration
        at src/TextShow/FromStringTextShow.hs:358:10-47
      or from: Show a
        bound by a quantified context
        at src/TextShow/FromStringTextShow.hs:1:1
      Possible fix:
        add (TextShow a) to the context of a quantified context
    • In the instance declaration for ‘Show1 (FromTextShow1 f)’
    |
358 | instance TextShow1 f => Show1 (FromTextShow1 f) where
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

And as @Ericson2314 notes in https://github.com/haskell/core-libraries-committee/issues/10#issuecomment-1166567641, I wouldn't be happy with putting `forall a. Show a => TextShow a` in the instance context, as that almost assuredly wouldn't work if you tried using it in practice.

In short, I'm stuck. Any ideas?

-----

>     1. Delete `Show1 FromTextShow` for being incorrect now, because having a `Show a` is _not_ sufficient to have a `Show (FromTextShow a)`: one instead needs a `TextShow a` and that is indeed the very point of this class.

We may well need to do this, but it's worth noting that this is a separate issue from the `FromTextShow1` instance issue I describe above. I genuinely have no clue how to fix that one.

**Comment by Ericson2314** [comment]
@RyanGlScott `FromTextShow1 f :.: FromStringShow` and `FromStringShow1 f :.: FromTextShow` *do* morally have `Show1` and `TextShow1` instances.

I suggested we

1. Add the quantified superclass constraints to the `*1` and `*2` classes.
2. Remove the now-broken instances.
3. Create new adopters to mimick the above instances we cannot actually write.

FWIW, note that the new adapters vaguely resemble profunctors in that one is mapping the instance one way before, and then mapping it back after.

**Comment by RyanGlScott** [comment]
I'm afraid I don't quite understand the plan you're proposing. Can you describe your plan in more detail, particularly steps (2) and (3)?

For what it's worth, the only `FromTextShow1` instance that I really care about is its `Show1` instance—I'd _really_ prefer not to change that if possible. Its `Show` instance is mostly there for the sake of completeness, and I'm happy to rearrange it as needed to make things typecheck.

**Comment by Ericson2314** [comment]
@RyanGlScott Ah sorry I was talking about the instances on the *non*-`*1` newtypes.

I'll give the `*1` newtypes some thought too, but also I think I better actually try out what ever sort of build you both are using so I can get some concrete errors.

**Comment by Bodigrim** [comment]
In the mean time, @Ericson2314, could you please rebase https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4727?

**Comment by Ericson2314** [comment]
Sure!

**Comment by RyanGlScott** [comment]
Thanks, @Ericson2314. FWIW, I've been prototyping things by simply redefining `Show1` and `Show2` locally with the quantified superclasses, which is far easier than building all of GHC.

**Comment by Ericson2314** [comment]
@RyanGlScott OK so I think we need 
```diff
-instance (TextShow1 f, TextShow a) => Show (FromTextShow1 f a)
+instance (TextShow1 f, Show a) => Show (FromTextShow1 f a)
```
And then `Show1 (FromTextShow1 f)` can be written again.

The new `Show` instance would do `(_ :: TextShowImpl -> ShowImpol) . TextShow1 f . (_ :: ShowImpl -> TextShowImpl)`  if that makes any sense.

To accomplish that I think we can change the definition of `FromTextShow1` to help (I'll write that next).

**Comment by Ericson2314** [comment]
```diff
-newtype FromTextShow1 f a = FromTextShow1 { fromTextShow1 :: f a }
+newtype FromTextShow1 f a = FromTextShow1
+  { fromTextShow1 :: FromTextShow (f (FromStringShow a))
+  }
```
and likewise
```diff
-newtype FromStringShow1 f a = FromStringShow1 { fromStringShow1 :: f a }
+newtype FromStringShow1 f a = FromStringShow1
+  { fromStringShow1 :: FromStringShow (f (FromTextShow a))
+  }
```

**Comment by Bodigrim** [comment]
I completed the impact assessment. The only Stackage packages which need patches are:
```cabal
source-repository-package
  type: git
  location: https://github.com/Bodigrim/kan-extensions.git

source-repository-package
  type: git
  location: https://github.com/Bodigrim/text-show.git

source-repository-package
  type: git
  location: https://github.com/Bodigrim/streaming.git

source-repository-package
  type: git
  location: https://github.com/Bodigrim/nonemptymap.git

source-repository-package
  type: git
  location: https://github.com/Bodigrim/sexp-grammar.git
  subdir: sexp-grammar

source-repository-package
  type: git
  location: https://github.com/Bodigrim/sandwich.git
  subdir: sandwich
```
Of which everything except `text-show` looks pretty straightforward. That said, the future progress on the proposal depends on sorting out a strategy for `text-show`.

**Comment by Ericson2314** [comment]
I am working on `text-show`.

**Comment by RyanGlScott** [comment]
> @RyanGlScott OK so I think we need
> 
> ```diff
> -instance (TextShow1 f, TextShow a) => Show (FromTextShow1 f a)
> +instance (TextShow1 f, Show a) => Show (FromTextShow1 f a)
> ```
> 
> And then `Show1 (FromTextShow1 f)` can be written again.

Sounds good to me. It's a bit of a shame that we have to do this solely for the sake of satisfying the superclass constraints, but at the same time, I don't have any better ideas.

> The new `Show` instance would do `(_ :: TextShowImpl -> ShowImpol) . TextShow1 f . (_ :: ShowImpl -> TextShowImpl)` if that makes any sense.
> 
> To accomplish that I think we can change the definition of `FromTextShow1` to help (I'll write that next).

Er, why do we need to do this? Wouldn't it suffice to just implement the `Show` instance as `showsPrec = showsPrec1`?

**Comment by Ericson2314** [comment]
See https://github.com/RyanGlScott/text-show/pull/55. Unfortunately I am scrambling to move so I don't think I will have time to work on it any more for the next few days, but it should hopefully make some sense in the interim?

> > To accomplish that I think e can change the definition of FromTextShow1 to help (I'll write that next).
>
> Er, why do we need to do this?

I thought it would be nice to keep track in the types of all the instances that need converting, But I am not sure I did that well, and the fact that `F` might not be representational is annoying.

On the other hand, it my allow avoiding the `*1`/`*2` classes by making the conversion canonical. That in the spirit of this issue, which has a goal of the quantified constraints to replacing those classes as much as possible.

**Comment by RyanGlScott** [comment]
I'm leery of making a breaking change to the definition of `FromTextShow1` when a non-breaking change will do. I'll put up a proof-of-concept MR soon, based the instance context idea proposed in https://github.com/haskell/core-libraries-committee/issues/10#issuecomment-1166584371. (My apologies for pushing back a bit when you clearly put a lot of work into RyanGlScott/text-show#55, but I do think this can be made quite a bit simpler.)

**Comment by Ericson2314** [comment]
@RyanGlScott Yeah sure that is fine, happy to hand it off. It is possible my PR is still an easier starting point to do that too but make the PR whoever you think is best.

**Comment by Bodigrim** [comment]
Dear CLC members, let's vote on https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4727. The proposal fixes an issue with instances for `Sum`, `Product` and `Compose` being too rigid, and establishes a relation between `Eq` and `Eq1` constraints and similar.

This is a breaking change, but the impact assessment (based on Stackage snapshot `nightly-2022-06-17`) shows only several packages affected, and we have PRs in place for all of them:
* https://github.com/RyanGlScott/text-show/pull/57
* https://github.com/haskell-streaming/streaming/pull/113
* https://github.com/ChristopherDavenport/nonemptymap/pull/7
* https://github.com/codedownio/sandwich/pull/53
* https://github.com/esmolanka/sexp-grammar/pull/26

Breakage typically stems from defining `instance Eq1 f` without `instance Eq a => Eq (f a)`. This can be fixed uniformly, by defining `instance Eq a => Eq (f a) where (==) = liftEq (==)`.

@tomjaguarpaw @chessai @cgibbard @emilypi @mixphix 

----

+1 from me, primarily for the reasons described in https://github.com/haskell/core-libraries-committee/issues/10#issuecomment-1166518746

**Comment by mixphix** [comment]
+1

**Comment by Ericson2314** [comment]
Note I am updating the MR with the thing that @phadej mentioned in the GHC MR thread. Amended the description of this issue accordingly.

**Comment by Ericson2314** [comment]
OK the PR is now up to date. The new revised `*2` constraints I think should be uncontroversial. The default methods might be controversial and if so I am happy to remove them.

**Comment by tomjaguarpaw** [comment]
The [MR text](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4727) is a bit odd. It starts off talking about `Eq`, then switches to `Show` before mentioning `Ord`. Was that intentional, for example to "mention the full range of changen"? I find it quite hard to read.

**Comment by Ericson2314** [comment]
Sorry, I didn't mean to derail the process.

The change made dates back a year to the conversation between @phadej and I in https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4727#note_357040 and is something that I had meant to add earlier but forgot about. I believe that change is very much in the spirit of the original (merely making the relationship between `*2` and `*1` match that between `*1` and `*`) and so it felt more like a small correction to fix an oversight and complete what the proposal started, rather than dragging the MR in a new direction.

I looked over the packages that needed to be changed so far (in hopes that it would be an interesting subset) and I believe that all would continue to work with the existing (pre this last change) fixes. Indeed https://github.com/ChristopherDavenport/nonemptymap/pull/7/files already does the `*1` classes in terms of the `*2` classes.

If you confirm you still want me to revert it I will, but I would propose the reverted part as a follow-up change right after. And if we do want to do both, I would suggest getting them both done in the same release to "get the breakage over with".

**Comment by Ericson2314** [comment]
OK while the the new super-classes I think are just better, I have no problem with holding off on the default methods since they are easy to add later but hard to remove. Keeping the latest constraints on the `*2` classes but reverting the default methodsis a fine option too.

**Comment by Bodigrim** [comment]
@Ericson2314 could you please clarify the MR description with respect to issues mentioned in https://github.com/haskell/core-libraries-committee/issues/10#issuecomment-1167839334 ? Once ready, please sign off that you are happy with the final version of the proposal to be put forward.

**Comment by Bodigrim** [comment]
@Ericson2314 could we please make some progress here?

**Comment by Ericson2314** [comment]
I am sorry for being slow/absent.

I have:

1. Fixed the commit message as requested.
2. Deleted the MR description (so there is one less thing to keep in sync, new description says look at commit message.) (Keep in mind that since GHC commits are rebased onto `master`, only the commit messages not MR description will live in version control.)
3. Updated this issue's description to match the MR.

I think that is all that is asked of me, but if there is anything more do let me know. I will try to be properly on top of this.

**Comment by Bodigrim** [comment]
@tomjaguarpaw are you satisfied with the updates?

**Comment by tomjaguarpaw** [comment]
+1

----

Yes, all good by me, thanks.

**Comment by Bodigrim** [comment]
Dear CLC members, let's resume the vote, started in https://github.com/haskell/core-libraries-committee/issues/10#issuecomment-1166669613. CC @chessai @cgibbard @emilypi 

@mixphix could you please confirm that your vote https://github.com/haskell/core-libraries-committee/issues/10#issuecomment-1166676009 is still valid?

**Comment by mixphix** [comment]
Yes, I'm still in favour. Thanks @Ericson2314 for the clarifications!

**Comment by cgibbard** [comment]
I'm still in favour as well. +1

**Comment by emilypi** [comment]
+1

**Comment by Bodigrim** [comment]
This gives us at least 5 votes in favor. Thanks all, approved.

**Comment by Bodigrim** [comment]
@Ericson2314 could you please prepare a migration guide for this change similar to https://github.com/haskell/core-libraries-committee/tree/main/guides ?

**Comment by Ericson2314** [comment]
Opened https://github.com/haskell/core-libraries-committee/pull/84 with the migration guide.

**Comment by mpickering** [comment]
It appears there are some packages on stackage which need changes due to this proposal but weren't flagged in the impact assessment. Do you know why they were missed?

For example: 

`dual`:

```
Control/Category/Dual.hs:23:10: error: [GHC-39999]
    • Could not deduce ‘Eq1 (Dual k a)’
        arising from the head of a quantified constraint
        arising from the superclasses of an instance declaration
      from the context: Eq2 k
        bound by the instance declaration
        at Control/Category/Dual.hs:23:10-30
      or from: Eq a
        bound by a quantified context at Control/Category/Dual.hs:23:10-30
    • In the instance declaration for ‘Eq2 (Dual k)’
   |
23 | instance Eq2 k => Eq2 (Dual k) where liftEq2 f g (Dual x) (Dual y) = liftEq2 g f x y
```

`selective`:

```
src/Control/Selective/Trans/Except.hs:51:7: error: [GHC-39999]
    • Could not deduce ‘Selective m’
        arising from the head of a quantified constraint
        arising from the 'deriving' clause of a data type declaration
      from the context: Monad m
        bound by a quantified context
        at src/Control/Selective/Trans/Except.hs:51:7-16
      Possible fix:
        add (Selective m) to the context of a quantified context
    • When deriving the instance for (MonadTrans (ExceptT e))
   |
51 |     , MonadTrans, MonadFix, MonadFail, MonadZip, MonadIO, MonadPlus, Eq1, Ord1
   | 
```

`bencoding`
```
src/Data/BEncode.hs:671:35: error: [GHC-39999]
    • No instance for ‘MonadPlus (Either String)’
        arising from the 'deriving' clause of a data type declaration
      Possible fix:
        use a standalone 'deriving instance' declaration,
          so you can specify the instance context yourself
    • When deriving the instance for (Alternative Get)
    |
671 |   deriving (Functor, Applicative, Alternative)
    |    
```

`either-both`:

```
Data/Either/Both.hs:17:10: error: [GHC-39999]
    • No instance for ‘Functor (Either' a)’
        arising from the head of a quantified constraint
        arising from the superclasses of an instance declaration
    • In the instance declaration for ‘Bifunctor Either'’
   |
17 | instance Bifunctor Either' where
   |          ^^^^^^^^^^^^^^^^^
```

`quickcheck-state-machine`:

```
[ 5 of 21] Compiling Test.StateMachine.Types.References ( src/Test/StateMachine/Types/References.hs, dist/
build/Test/StateMachine/Types/References.o, dist/build/Test/StateMachine/Types/References.dyn_o )

src/Test/StateMachine/Types/References.hs:94:10: error: [GHC-39999]
    • Could not deduce ‘Eq (Concrete a)’
        arising from the head of a quantified constraint
        arising from the superclasses of an instance declaration
      from the context: Eq a
        bound by a quantified context
        at src/Test/StateMachine/Types/References.hs:94:10-21
    • In the instance declaration for ‘Eq1 Concrete’
   |
94 | instance Eq1 Concrete where
   |   
```

**Comment by phadej** [comment]
> Do you know why they were missed?

Because impact assessment is done in june of 2022?

---

In case of `dual` the fix is straight-forward, yet maybe not that satisfactory:

```diff
--- a/Control/Category/Dual.hs
+++ b/Control/Category/Dual.hs
@@ -1,7 +1,7 @@
 {-# LANGUAGE DerivingVia #-}
 module Control.Category.Dual where

-import Prelude (Eq, Ord, Read, Show, Bounded, ($))
+import Prelude (Eq ((==)), Ord (compare), Read, Show, Bounded, ($))

 import Control.Category
 import Data.Bifunctor
@@ -20,6 +20,9 @@ instance Category k => Category (Dual k) where
     id = Dual id
     Dual f . Dual g = Dual (g . f)

+instance (Eq2 k, Eq a) => Eq1 (Dual k a) where liftEq f (Dual x) (Dual y) = liftEq2 f (==) x y
+instance (Ord2 k, Ord a) => Ord1 (Dual k a) where liftCompare f (Dual x) (Dual y) = liftCompare2 f compare x y
+
 instance Eq2 k => Eq2 (Dual k) where liftEq2 f g (Dual x) (Dual y) = liftEq2 g f x y
 instance Ord2 k => Ord2 (Dual k) where liftCompare2 f g (Dual x) (Dual y) = liftCompare2 g f x y

@@ -31,6 +34,9 @@ instance Show2 k => Show2 (Dual k) where
     liftShowsPrec2 asp asl bsp bsl n =
         showsUnaryWith (liftShowsPrec2 bsp bsl asp asl) "Dual" n . dual

+instance Bifunctor k => Functor (Dual k a) where
+    fmap f = Dual . bimap f id . dual
+
 instance Bifunctor k => Bifunctor (Dual k) where
     bimap f g = Dual . bimap g f . dual
 ```

 The problem is that there is no classes like

 ```haskell
 class FunctorOn2 p where
    anotherFirst :: (a -> b) -> p a x -> p b x
 ```

 so constraints for missing superclasses (`Eq1`, ..., `Functor`) will have overly restricted constraints. But at least we can write them, so it's not a blocker.

For `quickchekc-state-machine` the fix is

```
diff --git a/src/Test/StateMachine/Types/References.hs b/src/Test/StateMachine/Types/References.hs
index 9e4200d..89b59da 100644
--- a/src/Test/StateMachine/Types/References.hs
+++ b/src/Test/StateMachine/Types/References.hs
@@ -81,6 +81,8 @@ instance Ord1 Symbolic where
 data Concrete a where
   Concrete :: Typeable a => a -> Concrete a

+deriving stock instance Eq a => Eq (Concrete a)
+deriving stock instance Ord a => Ord (Concrete a)
 deriving stock instance Show a => Show (Concrete a)

 instance Show1 Concrete where
```

---

Especially the latter case is a problem of rolling stone uphill: without enforcement to have `Eq a => Foo (Eq a)` when you write `Eq1 Foo` instances, they will be omitted (cause humans makes mistakes). No different than `Generically1` case https://github.com/haskell/core-libraries-committee/issues/10#issuecomment-955640705, these just happen until the change is released.

**Comment by Bodigrim** [comment]
* [`dual`](https://hackage.haskell.org/package/dual) is a deprecate package, not much point to patch. 

* `selective` failure does not seem related, it is probably caused by changes to superclasses of `MonadTrans` in `transformers-0.6`.

* `bencoding` is again unrelated, it's just that `transformers-0.6` stopped providing `instance MonadPlus (Either String)`.

* [`either-both`](https://hackage.haskell.org/package/either-both) is again deprecated.

* `quickcheck-state-machines` was not a part of [Stackage Nightly 2022-06-17](https://www.stackage.org/nightly-2022-06-17/cabal.config) which was used for impact assessment.

**Comment by chshersh** [comment]
I'm trying to summarise the state of this proposal as part of my volunteering effort to track the progress of all `approved` CLC proposals.

| Field | Value |
|---|---|
| **Author**             |  @Ericson2314   |
| **Status**             |  merged |
| **`base` version**     | `4.18.0.0`  |
| **Merge Request (MR)** | https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4727 |
| **Blocked by**         | nothing |
| **CHANGELOG entry**    | present |
| **Migration guide**    | https://github.com/haskell/core-libraries-committee/blob/main/guides/functor-combinator-instances-and-class1s.md |

[changelog]: https://gitlab.haskell.org/ghc/ghc/-/blob/ghc-9.6/libraries/base/changelog.md

Please, let me know if you find any mistakes 🙂
